packages/flutter_tools/lib/src/test/runner.dart

```diff
@@ -89,14 +123,65 @@ Future<int> runTests(
 
   // Call package:test's main method in the appropriate directory.
   final Directory saved = fs.currentDirectory;
+
+  final Function printReloadMessage = () {
+    printStatus(
+      "Press 'r' to rerun your tests, 'q' to quit",
+      color: TerminalColor.red
+    );
+  };
+
   try {
     if (workDir != null) {
       printTrace('switching to directory $workDir to run tests');
       fs.currentDirectory = workDir;
     }
 
     printTrace('running test package with arguments: $testArgs');
-    await test.main(testArgs);
+    await test.runTests(testArgs);
+
+    if (watchTests) {
+      final Completer<void> completer = Completer<void>();
+      final DirectoryWatcher directoryWatcher = DirectoryWatcher(saved.path);
+
+      directoryWatcher.events.listen((WatchEvent event) {
+        if (event.path.endsWith('.dart')) {
+          invalidatedFiles.add(event.path);
+        }
+      });
+
+      terminal.singleCharMode = true;
+      terminal.keystrokes.asBroadcastStream().listen((String char) async {
```

Comment:
I would not structure this with a completer + listeners, instead use `async` and `await for`. I also would not use a switch with 2 cases, instead just use if.

Example:
```
// setup
await directoryWatcher.ready;
await for (String char in terminal.keystrokes.asBroadcastStream()) {
  if (char == 'r') {

  } else if (char == 'q') {
    break;
  }
}

```

Reply:
Await for is much nicer, thanks for that.  The switch is here to make it nicer for anything else we'd like to add, but as that's unlikely I guess that's a premature optimization.  Will fix.

Reply:
Done